File: core/src/vct/col/ast/TypeExpression.java
package vct.col.ast;

import java.util.Arrays;

public class TypeExpression extends Type {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  public final TypeOperator op;
  public final Type types[];
  
  public TypeExpression(TypeOperator op,Type ... t) {
    this.op=op;
    types=Arrays.copyOf(t,t.length);
  }
  
  public boolean isNumeric() {
    switch(op){
    case Local:
    case Global:
    case Long:
      return types[0].isNumeric();
    default:
      return false;
    }
  }

  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
 
  @Override
  public <T> T accept_simple(TypeMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }


  @Override
  public boolean supertypeof(ProgramUnit context, Type t) {
    return false;
  }

}

File: vercors/src/main/java/vct/col/ast/type/TypeExpression.scala
package vct.col.ast.`type`

import scala.collection.JavaConverters._
import vct.col.ast.stmt.decl.ProgramUnit
import vct.col.ast.util.{ASTVisitor, TypeMapping}
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}

case class TypeExpression(val operator:TypeOperator, val types:List[Type]) extends Type with VisitorHelper {
  require(types != null, "The types list is null")

  /** Constructs a new type expression from an array of types */
  def this(operator:TypeOperator, types:Array[Type]) = this(operator, types.toList)

  /** Gives the heading type in the type list */
  def firstType = types.head

  /** Provides a Java wrapper (as `java.util.List`) over the types list (`types`) */
  def typesJava = types.asJava

  override def isNumeric = operator match {
    case TypeOperator.Local | TypeOperator.Global | TypeOperator.Long => firstType.isNumeric
    case _ => false
  }

  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
  override def accept_simple[T](m:TypeMapping[T]) = handle_standard(() => m.map(this))
  override def supertypeof(context:ProgramUnit, t:Type) = false

  override def debugTreeChildrenFields: Iterable[String] = Seq("types", "args")
  override def debugTreePropertyFields: Iterable[String] = Seq("operator")
}
